### PR TITLE
Add Face Sign functionality

### PR DESCRIPTION
**Pull request by Kink3d**
# Add Face Sign functionality to all subshaders

## Purpose

Adds ability for change graph output depending on the face sign of a given fragment.

## Changes

- Add `FaceSignNode` (returns true if front face)
- Add `IMayRequireFaceSign` interface
- Add `FaceSign` requirement calculations
- Add `FaceSign` subscript to `SurfaceDescriptionInputs` struct in HD templates
- Add `FaceSign` optional dependencies to HD subshaders
- Add `FaceSign` subscript to `SurfaceDescriptionInput` calculation in LW and preview subshaders
- Add `VFACE` semantic to fragment argument in LW and preview templates
- Add `FaceSign` to `ShaderGeneratorNames`
- Fix reflection in GraphUtil attempting `float` (changed to `Single`)

## Risk

- Changes to HD and LW templates
- Addition to general requirement calculations

### CODE REVIEW

**Comment by Katstodian** [comment]
My review of the changelog.md will have to wait until Peter's comment is resolved. Changing the name of the node will affect the documentation. :)

**Comment by Kink3d** [comment]
Ive changed the slot name to `Is Front` to make the output value clearer. As discussed during our meeting earlier the node name will not change as it is an expected name amongst artists.

File: com.unity.shadergraph/Editor/Data/Util/GraphUtil.cs [hunk 1]
```diff
@@ -92,7 +92,7 @@ private static int GetFloatVectorCount(string typeName)
             {
                 return 2;
             }
-            else if (typeName.Equals("float"))
+            else if (typeName.Equals("Single"))
```
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Data/Util/GraphUtil.cs]: this won't break anyone using 'float' ?

File: com.unity.shadergraph/CHANGELOG.md [hunk 2]
```diff
@@ -21,6 +21,11 @@ In the expanded property window, you can now also toggle if the property is expo
 
 You can now change the path of Shader Graphs and Sub Graphs. When you change the path of a Shader Graph, this modifies the location it has in the shader selection list. When you change the path of Sub Graph, it will have a different location in the node creation menu.
 
+### Face Sign node
+
+![](.data/face_sign.png)
+
+This node adds the ability to change graph output depending on the face sign of a given fragment. The Face Sign node returns true is currently rendering a front face, and false if rendering a back face. Enable "two sided" on the Master node to use this functionality.
```
Comment (Katstodian) [inline com.unity.shadergraph/CHANGELOG.md]: Please replace this:
This node adds the ability to change graph output depending on the face sign of a given fragment. The Face Sign node returns true is currently rendering a front face, and false if rendering a back face. Enable "two sided" on the Master node to use this functionality.

With this:
With this node, you can change graph output depending on the face sign of a given fragment. To render a front face, set the node to True. To render a back face, set it to False. 
Note: This functionality requires that **two sided** is enabled on the Master node.

Comment (pbbastian) [reply com.unity.shadergraph/CHANGELOG.md]: Hmm, this makes it sound like you're setting the value _for_ the node, but actually the node is returning a value _to_ you, which is true or false depending on whether the current fragment being rendered is part of a front or back face.

Comment (Katstodian) [reply com.unity.shadergraph/CHANGELOG.md]: Ahh, I see your point. I'll edit :)

Comment (Katstodian) [reply com.unity.shadergraph/CHANGELOG.md]: Updated. @pbbastian, technical approval?

File: com.unity.shadergraph/Editor/Data/Util/GraphUtil.cs (hunk 1 above)
Comment (Kink3d) [reply com.unity.shadergraph/Editor/Data/Util/GraphUtil.cs]: This wasnt used anywhere else in the code that I could find. I am using a float, Reflection returns it as `Single`

File: com.unity.shadergraph/CHANGELOG.md (hunk 2 above)
Comment (pbbastian) [reply com.unity.shadergraph/CHANGELOG.md]: Yes 👍

**Comment by zadi0648** [comment]
Caught a bug with textures leaking through the back side, Matt fixed it, approved.

**Pull request merged**